docs/closure-compiler.md

```diff
@@ -0,0 +1,697 @@
+ *   wasElementMadeActive: boolean,
+ *   activationStartTime: number,
+ *   activationEvent: ?Event
+ * }}
+ */
+let ActivationStateType;
+
+/**
+ * @typedef {{foo: number}}
+ */
+export let MyExportedType;
+```
+
+While these `let` declarations do not do anything at runtime, they are used by closure to
+encapsulate complex types as specified through a [\@typedef statement](https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System#typedefs). The statements above let both `ActivationStateType` and `MyExportedType` be used as type
+parameters throughout the rest of the code.
+
+### "Dictionary" objects via quoted keys / bracket references
+
+**Example**:
+
+```js
+/** @dict */
+const SETTINGS = {
+  'numRetries': 1,
+  'selectorToQuery': 'body',
+  'windowObject': window,
+};
+
+window.settings = SETTINGS;
+
+/** @const {!Object<string, string>} **/
+const DEACTIVATION_ACTIVATION_PAIRS = {
+  'mouseup': 'mousedown',
+  'pointerup': 'pointerdown',
+  'touchend': 'touchstart',
+  'keyup': 'keydown',
+  'blur': 'focus',
+};
+
+Object.keys(DEACTIVATION_ACTIVATION_PAIRS).forEach((deactivationEvt) => {
+  const activationEvt = DEACTIVATION_ACTIVATION_PAIRS[deactivationEvt];
+  domNode.addEventListener(activationEvt, someActivationListener);
+  domNode.addEventListener(deactivationEvt, someDeactivationListener);
+});
+
+/** @const {!Array<!Object<string, !Function>>} */
+const listeners = [
+  {
+    'mouseup': () => console.log('mouseup'),
+    'mousedown': () => console.log('mousedown'),
+  },
+  {
+    'keyup': () => console.log('keyup'),
+    'keydown': () => console.log('keydown'),
+  },
+];
+
+console.log(listeners[0]['mouseup']);
+```
+
+Sometimes in our code, object keys will have meaning outside just being a key for an object. An
+example of this might be a map of event types to their respective listeners, as shown above.
+
+When closure compiles javascript using advanced optimizations, it [obfuscates property values](https://developers.google.com/closure/compiler/docs/api-tutorial3#better) in
+objects, classes, etc. While this leads to smaller code, it also leads to issues when those
+properties are used externally. In order to prevent this behavior, object keys need to be quoted,
+and those keys need to be referenced using bracket notation.
+
+### Member declarations via sentinel property reference expressions.
+
+Closure enforces that properties cannot be added to object instances unless those properties are
+specified in their constructors. This is problematic in cases where a base constructor calls a
+"setup" function in which properties are added to the instance. This is a pattern we use in our
+codebase via `MDCComponent#initialize()`, so that the component has an opportunity to perform any
+instantiation logic without losing all of the base constructor logic of assigning a root element,
+instantiating an adapter, etc. Because `initialize()` is called as part of the `super()` call within
```

Comment:
Also this might be a obvious question but I didn't quite get it.

What does closure mean property cannot be added outside constructor? It seems to me that `initialize` is called as part of `constructor`, so the property is added to `MDCComponent` at that time. So I wonder if that has to do with how closure identify Javascript is adding property? Does it mean literally that the property need to be defined within `constructor`?